src/NHapi.Base/Llp/CharSetUtility.cs

```diff
@@ -0,0 +1,150 @@
+                {
+                    encoding = Hl7CharSets.FromHl7Encoding(hl7CharsetName);
+                }
+
+                Log.Trace($"Detected MSH-18 value {hl7CharsetName} so using encoding {encoding.EncodingName}");
+            }
+            catch (EncodingNotSupportedException ex)
+            {
+                Log.Warn($"Invalid or unsupported encoding in MSH-18. Defaulting to {encoding.EncodingName}", ex);
+            }
+            catch (HL7Exception ex)
+            {
+                Log.Warn($"Failed to parse MSH segment. Defaulting to {encoding.EncodingName}", ex);
+            }
+
+            return encoding;
+        }
+
+        private static string StripNonLowAscii(string theString)
+        {
+            if (theString == null)
+            {
+                return string.Empty;
+            }
+
+            var builder = new StringBuilder();
+
+            foreach (var next in theString)
+            {
+                if (next > 0 && next < 127)
+                {
+                    builder.Append(next);
+                }
+            }
+
+            return builder.ToString();
+        }
+
+        private class Bom
+        {
+            private static readonly IList<Bom> KnownBoms = new List<Bom>
+            {
+                new Bom(new byte[] { }, Encoding.ASCII),
+                new Bom(new byte[] { 0xEF, 0xBB, 0xBF }, Encoding.UTF8),
+                new Bom(new byte[] { 0xFF, 0xFE }, Encoding.Unicode),                           // UTF-16LE
+                new Bom(new byte[] { 0xFE, 0xFF, 0xBF }, Encoding.BigEndianUnicode),            // UTF-16BE
+                new Bom(new byte[] { 0xFF, 0xFE, 0x00, 0x00 }, Encoding.UTF32),                 // UTF-32LE
+                new Bom(new byte[] { 0x00, 0x00, 0xFE, 0xFF }, new UTF32Encoding(true, true)),  // UTF-32BE
+            };
+
+            public Bom(byte[] bomBytes, Encoding encoding)
+            {
+                BomBytes = bomBytes;
+                Encoding = encoding;
+            }
+
+            public byte[] BomBytes { get; }
+
+            public Encoding Encoding { get; }
+
+            public static string SkipBom(byte[] messageBytes)
+            {
+                var bom = GetBom(messageBytes);
+                var encoding = bom.Encoding;
+                var messageBytesWithoutBom = messageBytes.Skip(bom.BomBytes.Length).ToArray();
+                return encoding.GetString(messageBytesWithoutBom);
+            }
+
+            public static Bom GetBom(byte[] messageBytes)
+            {
+                if (messageBytes == null)
+                {
+                    return KnownBoms[0];
+                }
+
+                foreach (var bom in KnownBoms)
+                {
+                    var messageBomBytes = messageBytes.Take(bom.BomBytes.Length);
+                    if (bom.BomBytes.SequenceEqual(messageBomBytes))
+                    {
+                        return bom;
```

Comment:
I think this will always return ASCII BOM, because in the KnownBoms it is first and it has 0 bytes, so .Take will take 0 bytes and empty array with empty array will be true with SequenceEqual. You probably want to foreach through KnownBoms without ASCII BOM in here.

Reply:
@jakubsuchybio will have a look at this when I get chance.
not an easy one to unit test without mocking messages with leading boms 😢 not too hard I suppose just a pain to setup.

Reply:
I would just test the `GetBom` method, very small unit test that tests, when you have some non-empty BOM, then that you get that bom and not the ASCII

Reply:
@jakubsuchybio I have moved the empty bom to be the last item in the list, the same as in `hapi` now since lists are iterated in natural order i.e. `KnownBoms[0], KnownBoms[1], KnownBoms[2]` etc.

by moving it to the end it will only use that as a fall back.

I will still try to find time to add some tests though.

Reply:
@jakubsuchybio okay have fixed this, added tests and pushed trhe changes.